docs: add environment addons documentation (#4378)
aws/copilot-cli: site/content/docs/developing/addons/environment.en.md

# Modeling Additional Environment Resources with AWS CloudFormation

Additional AWS resources, referred to as "addons" in the CLI, are any additional AWS services that a [service or environment manifest](../../manifest/overview.en.md) does not integrate by default. 
For example, an addon can be a DynamoDB table, an S3 bucket, or an RDS Aurora Serverless cluster that your service needs to read or write to.

You can define additional resources for a workload (such as a [Load Balanced Web Service](../../manifest/lb-web-service.en.md)
or a [Scheduled Job](../../manifest/scheduled-job.en.md)).

> The lifecycle of workload addons will be managed by the workload and will be deleted once the workload is deleted.

Reviewer: Should we remove this blurb from the environment addons page?

Author: My thought was to include the same high-level description of addons for both workload and environment page, and then say "this page documents workload/env addons". Otherwise, for example if a user lands on the workload addon page, they might think oh addon is for workload-level only without noticing there is another "additional environment resource" page

Author: <img width="793" alt="Screen Shot 2023-01-17 at 3 45 51 PM" src="https://user-images.githubusercontent.com/79273084/213019115-eaf519c4-78a0-44c5-8754-b4e1a3c754c9.png">

I updated the blog:
1. Kept the top-level description (lmk your thought!!)
2. Make it clearer that there is another doc for workload/environment addon in the environment/workload doc page, respectively

# Modeling Additional Environment Resources with AWS CloudFormation

Additional AWS resources, referred to as "addons" in the CLI, are any additional AWS services that a [service or environment manifest](../../manifest/overview.en.md) does not integrate by default. 
For example, an addon can be a DynamoDB table, an S3 bucket, or an RDS Aurora Serverless cluster that your service needs to read or write to.

You can define additional resources for a workload (such as a [Load Balanced Web Service](../../manifest/lb-web-service.en.md)
or a [Scheduled Job](../../manifest/scheduled-job.en.md)).
The lifecycle of workload addons will be managed by the workload and will be deleted once the workload is deleted.

Alternatively, you can define additional shareable resource for an environment.
Environment addons won't be deleted unless the environment is deleted.

This page documents how to create environment-level addons.

## How do I add an S3 bucket, a DDB Table, or an Aurora Serverless cluster?

!!!caution "" 

>     As of v1.25.0, `copilot storage init` does not supports creating addons on environment-level.

Reviewer 2: ```suggestion
    As of v1.25.0, `copilot storage init` does not support creating addons at the environment level.
```

## How to add other resources?

For other types of addons, you can add your own custom CloudFormation templates:

1. You can store the custom templates in your workspace under `copilot/environments/addons` directory.
2. When running `copilot env deploy`, the custom addons template will be deployed along with your environment stack.

???- note "Sample workspace layout with environment addons"
    ```term
    .
    └── copilot
        └── environments
            ├── addons  # Store environment addons.
            │   └── mys3.yaml
            ├── dev
            └── prod      
    ```

## What does an addon template look like?
An environment addon template can be [any valid CloudFormation template](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/template-anatomy.html) that satisfies the following:

> * Include at least one [`Resource`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/resources-section-structure.html).

Reviewer 2: ```suggestion
* Includes at least one [`Resource`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/resources-section-structure.html).
```

> * The `Parameters` section must include `App`, `Env`.

Reviewer 2: Because this is an item in the list of "the following"

```suggestion
* The `Parameters` section includes `App`, `Env`.
```

            Export:
                Name: !Sub ${App}-${Env}-MyTableARN  # This value can be consumed by a workload manifest or a workload addon.
        MyTableName:
            Value: !Ref ServiceTable
            Export:
                Name: !Sub ${App}-${Env}-MyTableName
    ```

It is important that you add the `Export` block. 
Otherwise, your workload stack or your workload addons won't be able to access the value.
You will use `Export.Name` to reference the value from your workload-level resources.

???- hint "Consideration: Namespace your `Export.Name`"
    You can specify any name you like for `Export.Name`.
    That is, it doesn't have to be prefixed with `!Sub ${App}-${Env}`; it can simply be `MyTableName`.

    However, within an AWS region, the `Export.Name` must be unique. 
    That is, you can't have two exports named `MyTableName` in `us-east-1`.
    
    Therefore, we recommend you to namespace your exports with `${App}` and `${Env}` to decrease the chance of name collision. 
    In addition, this makes it clear which application and environment the value is managed under.
    
    With the namespace, for example, say your application's name is `"my-app"`,

>     and you deployed the addons with environment `test`, then the final export name would be rendered to  `my-app-test-MyTableName`.

Reviewer 2: ```suggestion
    and you deployed the addons with environment `test`, then the final export name would be rendered to `my-app-test-MyTableName`.
```

##### Referencing from a workload addon

In your workload addons, you can reference a value from your environment addons, as long as that value is exported.

> To do so, use [`Fn::ImportValue`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-importvalue.html)

Reviewer 2: ```suggestion
To do so, use the [`Fn::ImportValue`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-importvalue.html)
```

function with that value's export name to import it from an environment addon.

???- note "Example: An IAM policy to access an environment-level DynamoDB table"
    ```yaml
    Parameters:
      App:
        Type: String
        Description: Your application's name.
      Env:
        Type: String
        Description: The environment name your service, job, or workflow is being deployed to.
      Name:
        Type: String
        Description: The name of the service, job, or workflow being deployed.
    Resources:
      MyTableAccessPolicy:
        Type: AWS::IAM::ManagedPolicy
        Properties:
          Description: Grants CRUD access to the Dynamo DB table
          PolicyDocument:
            Version: '2012-10-17'
            Statement:
              - Sid: DDBActions
                Effect: Allow
                Action:

>                   - dynamodb:* # NOTE: Scope down the permissions in your real application. This is done so that the blog isn't too long!

Reviewer 2: ```suggestion
                  - dynamodb:* # NOTE: Scope down the permissions in your real application. This is done so that this example isn't too long!
```
